File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/fragments/SoftwareVersionsFragment.java
package de.dimedis.mobileentry.fragments;

import android.annotation.SuppressLint;
import android.database.Cursor;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.loader.app.LoaderManager;
import androidx.loader.content.CursorLoader;
import androidx.loader.content.Loader;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import de.dimedis.mobileentry.R;
import de.dimedis.mobileentry.backend.response.DownloadLanguagesResponse;
import de.dimedis.mobileentry.backend.response.DownloadMyAvailableBordersResponse;
import de.dimedis.mobileentry.backend.response.DownloadOfflineConfigResponse;
import de.dimedis.mobileentry.backend.response.GetVersionsResponse;
import de.dimedis.mobileentry.backend.response.Versions;
import de.dimedis.mobileentry.databinding.FragmentSoftwareversionBinding;
import de.dimedis.mobileentry.databinding.SoftversionBinding;
import de.dimedis.mobileentry.db.content_provider.DataContentProvider;
import de.dimedis.mobileentry.service.HeartbeatService;
import de.dimedis.mobileentry.ui.InitActivity;
import de.dimedis.mobileentry.util.ConfigPrefHelper;
import de.dimedis.mobileentry.util.Logger;
import de.dimedis.mobileentry.util.UpdateUtil;
import de.dimedis.mobileentry.util.dynamicres.DynamicString;

public class SoftwareVersionsFragment extends ProgressLoaderFragment implements LoaderManager.LoaderCallbacks<Cursor> {
    static final String TAG = "SoftwareVersions";

    static final int LOADER_ID = 225;

    FragmentSoftwareversionBinding binding;
    SoftversionBinding softversionBinding;

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private final Bundle arguments;

        public Builder() {
            arguments = new Bundle();
        }

        public SoftwareVersionsFragment build() {
            SoftwareVersionsFragment fragment = new SoftwareVersionsFragment();
            fragment.setArguments(arguments);
            return fragment;
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
        UpdateUtil.updateSettings(() -> getMainActivity().runOnUiThread(() -> {
            checkUpdate();
            setUpdateButtonVisibility();
            setVisibilityProgressBar(false);
        }));
        HeartbeatService.scheduleService();
    }

    @Override
    public void onStop() {
        EventBus.getDefault().unregister(this);
        super.onStop();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        binding = FragmentSoftwareversionBinding.inflate(inflater, container, false);
        softversionBinding = binding.softVer;

        binding.startLogin.setOnClickListener(view -> {
            fragment(LoginScanFragment.builder().build());
        });

        getLoaderManager().initLoader(LOADER_ID, null, this);

        return binding.getRoot();
    }

    protected boolean isVersions;
    protected boolean isDownloadMyAvailableBorders;
    protected boolean isDownloadOfflineConfig;
    protected boolean isDownloadSettings;
    protected boolean isLocalization;

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(DownloadLanguagesResponse event) {
        responseProcess(event, new Response<DownloadLanguagesResponse>() {
            @Override
            public void resultOK(DownloadLanguagesResponse event) {
                Logger.i(SoftwareVersionsFragment.class.getSimpleName(), "got download lang response");
                getConfigPreferences().setIsLanguagesInit(true);
                DynamicString.getInstance().setLang(getConfigPreferences().currentLanguage());
                DynamicString.scan(getView());
                isLocalization = true;
                checkForStartLogin();
            }

            @Override
            public boolean onError(DownloadLanguagesResponse event) {
                Logger.i(SoftwareVersionsFragment.class.getSimpleName(), "got download lang response error");
                isLocalization = true;
                checkForStartLogin();
                return false;
            }
        });
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(GetVersionsResponse event) {
        Log.i(TAG, "onEvent GetVersionsResponse:" + event);
        responseProcess(event, new ResponseImpl<GetVersionsResponse>() {
            @Override
            public void resultOK(GetVersionsResponse event) {
                checkUpdate();
                setUpdateButtonVisibility();
                isVersions = true;
                checkForStartLogin();
            }
        });
    }

    //DownloadMyAvailableBordersResponse
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(DownloadMyAvailableBordersResponse event) {
        Log.i(TAG, "onEvent DownloadMyAvailableBordersResponse event:" + event);
        responseProcess(event, new ResponseImpl<DownloadMyAvailableBordersResponse>() {
            @Override
            public void resultOK(DownloadMyAvailableBordersResponse event) {
                isDownloadMyAvailableBorders = true;
                getConfigPreferences().setIsBordersInit(true);
                checkForStartLogin();
            }
        });
    }

    ///DownloadOfflineConfigResponse
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(DownloadOfflineConfigResponse event) {
        Log.i(TAG, "onEventMainThread DownloadOfflineConfigResponse event:" + event);
        responseProcess(event, new ResponseImpl<DownloadOfflineConfigResponse>() {
            @Override
            public void resultOK(DownloadOfflineConfigResponse event) {
                isDownloadOfflineConfig = true;
                checkForStartLogin();
            }
        });
    }

    void checkForStartLogin() {
        if (isVersions & getConfigPreferences().isBordersInit() & getConfigPreferences().isLanguagesInit() & isDownloadOfflineConfig & isDownloadSettings) {
            checkUpdate();
            setUpdateButtonVisibility();
            setVisibilityProgressBar(false);
        }
    }

    public void checkUpdate() {
        softversionBinding.textInfo.setVisibility(ConfigPrefHelper.isUpdateAvailable() ? View.VISIBLE : View.GONE);
        _fillVersionData();
    }

    protected void setUpdateButtonVisibility() {
    }

    boolean checkAndSetItem(TextView textView, String txt, TextView textViewServer, String txtServer, boolean isApp) {
        textView.setText(txt);
        if (isApp && txtServer == null) {
            textView.setText(getLocalizedString(R.string.SOFTWARE_VERSIONS_APP_VERSION_UNKNOWN_TEXT));
            textViewServer.setVisibility(View.GONE);
            return false;
        }
        if (!txtServer.equals(txt)) {
            String displayText = " " + txtServer; // Added space before txtServer
            textViewServer.setText(displayText);
            textViewServer.setVisibility(View.VISIBLE);

            if (txt.compareTo(txtServer) < 0) {
                textViewServer.setTextAppearance(R.style.TextSmallRed);
            } else {
                textViewServer.setTextAppearance(R.style.TextSmall);
            }
            return true;
        } else {
            textViewServer.setVisibility(View.GONE);
            return false;
        }
    }

    protected boolean isNeedUpdateApp = false;
    protected boolean isNeedUpdateLibrary = false;
    protected boolean isNeedUpdateLanguages = false;
    protected boolean isNeedUpdateBorders = false;
    protected boolean isNeedUpdateLocalConfig = false;

    void fillVersionsContext(Versions versions, Versions versionsServer) {


        isNeedUpdateApp = checkAndSetItem(softversionBinding.app, versions.getApp(), softversionBinding.appNew, versionsServer.getApp(), true);
        isNeedUpdateLibrary = checkAndSetItem(softversionBinding.lib, versions.getLibrary(), softversionBinding.libNew, versionsServer.getLibrary(), false);
        isNeedUpdateLanguages = checkAndSetItem(softversionBinding.lang, versions.getLanguages(), softversionBinding.langNew, versionsServer.getLanguages(), false);
        isNeedUpdateBorders = checkAndSetItem(softversionBinding.border, versions.getMyAvailableBorders(), softversionBinding.borderNew, versionsServer.getMyAvailableBorders(), false);
        isNeedUpdateLocalConfig = checkAndSetItem(softversionBinding.local, versions.getLocalConfig(), softversionBinding.localNew, versionsServer.getLocalConfig(), false);

        View v = getView();
        if (v != null) {
            v.invalidate();
        }
    }


    public void _fillVersionData() {
        fillVersionsContext(ConfigPrefHelper.getVersions(), ConfigPrefHelper.getVersionsFromServer());
    }

    @Override
    void onErrorCancelClick() {
        getActivity().finish();
        InitActivity.start(getContext());
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        return new CursorLoader(getContext(), DataContentProvider.TICKETS_URI, new String[]{"COUNT(*)"}, null, null, null);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor cursor) {
        int codeCount = 0;
        if (cursor != null && cursor.moveToNext()) {
            codeCount = cursor.getInt(0);
        }
        softversionBinding.softVer.setText(Integer.toString(codeCount));// + " " + codeCountText);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
    }

    @Override
    public void onDestroy() {  // TODO Deprecated
        getActivity().getSupportLoaderManager().destroyLoader(0);
        super.onDestroy();
    }
}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/fragments/SoftwareVersionsFragment.kt
package de.dimedis.mobileentry.fragments

import android.annotation.SuppressLint
import android.database.Cursor
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import de.dimedis.mobileentry.ConfigPref
import de.dimedis.mobileentry.R
import de.dimedis.mobileentry.backend.response.*
import de.dimedis.mobileentry.databinding.FragmentSoftwareversionBinding
import de.dimedis.mobileentry.databinding.SoftversionBinding
import de.dimedis.mobileentry.db.content_provider.DataContentProvider
import de.dimedis.mobileentry.service.HeartbeatService
import de.dimedis.mobileentry.ui.InitActivity
import de.dimedis.mobileentry.util.ConfigPrefHelper
import de.dimedis.mobileentry.util.ICompleteOperationCallback
import de.dimedis.mobileentry.util.Logger
import de.dimedis.mobileentry.util.UpdateUtil.updateSettings
import de.dimedis.mobileentry.util.dynamicres.DynamicString
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class SoftwareVersionsFragment : ProgressLoaderFragment(), LoaderManager.LoaderCallbacks<Cursor?> {
    var binding: FragmentSoftwareversionBinding? = null
    var softversionBinding: SoftversionBinding? = null

    class Builder {
        private val arguments: Bundle = Bundle()

        fun build(): SoftwareVersionsFragment {
            val fragment = SoftwareVersionsFragment()
            fragment.arguments = arguments
            return fragment
        }
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
        updateSettings(object : ICompleteOperationCallback {
            override fun onOperationComplete() {
                getMainActivity()!!.runOnUiThread {
                    checkUpdate()
                    setUpdateButtonVisibility()
                    setVisibilityProgressBar(false)
                }
            }
        })
        HeartbeatService.scheduleService()
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        super.onCreateView(inflater, container, savedInstanceState)
        binding = FragmentSoftwareversionBinding.inflate(inflater, container, false)
        softversionBinding = binding!!.softVer
        binding!!.startLogin.setOnClickListener {
            fragment(LoginScanFragment.builder().build())
        }
        loaderManager.initLoader(LOADER_ID, null, this)
        return binding!!.root
    }

    protected var isVersions = false
    protected var isDownloadMyAvailableBorders = false
    protected var isDownloadOfflineConfig = false
    protected var isDownloadSettings = false
    protected var isLocalization = false
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: DownloadLanguagesResponse?) {
        if (event != null) {
            responseProcess(event, object : Response<DownloadLanguagesResponse?> {
                override fun resultOK(event: DownloadLanguagesResponse?) {
                    Logger.i(SoftwareVersionsFragment::class.java.simpleName, "got download lang response")
                    ConfigPref.isLanguagesInit = (true)
                    DynamicString.instance.setLang(ConfigPref.currentLanguage)
                    DynamicString.scan(view)
                    isLocalization = true
                    checkForStartLogin()
                }

                override fun onError(event: DownloadLanguagesResponse?): Boolean {
                    Logger.i(SoftwareVersionsFragment::class.java.simpleName, "got download lang response error")
                    isLocalization = true
                    checkForStartLogin()
                    return false
                }
            })
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: GetVersionsResponse) {
        Log.i(TAG, "onEvent GetVersionsResponse:$event")
        responseProcess(event, object : ResponseImpl<GetVersionsResponse?>() {
            override fun resultOK(event: GetVersionsResponse?) {
                checkUpdate()
                setUpdateButtonVisibility()
                isVersions = true
                checkForStartLogin()
            }
        })
    }

    //DownloadMyAvailableBordersResponse
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: DownloadMyAvailableBordersResponse) {
        Log.i(TAG, "onEvent DownloadMyAvailableBordersResponse event:$event")
        responseProcess(event, object : ResponseImpl<DownloadMyAvailableBordersResponse?>() {
            override fun resultOK(event: DownloadMyAvailableBordersResponse?) {
                isDownloadMyAvailableBorders = true
                ConfigPref.isBordersInit = (true)
                checkForStartLogin()
            }
        })
    }

    ///DownloadOfflineConfigResponse
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: DownloadOfflineConfigResponse) {
        Log.i(TAG, "onEventMainThread DownloadOfflineConfigResponse event:$event")
        responseProcess(event, object : ResponseImpl<DownloadOfflineConfigResponse?>() {
            override fun resultOK(event: DownloadOfflineConfigResponse?) {
                isDownloadOfflineConfig = true
                checkForStartLogin()
            }
        })
    }

    fun checkForStartLogin() {
        if (isVersions and ConfigPref.isBordersInit and ConfigPref.isLanguagesInit and isDownloadOfflineConfig and isDownloadSettings) {
            checkUpdate()
            setUpdateButtonVisibility()
            setVisibilityProgressBar(false)
        }
    }

    fun checkUpdate() {
        softversionBinding!!.textInfo.visibility = if (ConfigPrefHelper.isUpdateAvailable()) View.VISIBLE else View.GONE
        _fillVersionData()
    }

    protected fun setUpdateButtonVisibility() {}
    fun checkAndSetItem(textView: TextView, txt: String, textViewServer: TextView, txtServer: String?, isApp: Boolean): Boolean {
        textView.text = txt
        if (isApp && txtServer == null) {
            textView.text = getLocalizedString(R.string.SOFTWARE_VERSIONS_APP_VERSION_UNKNOWN_TEXT)
            textViewServer.visibility = View.GONE
            return false
        }
        return if (txtServer != txt) {
            textViewServer.text = txtServer
            textViewServer.visibility = View.VISIBLE
            if (txt < txtServer!!) {
                textViewServer.setTextAppearance(R.style.TextSmallRed)
            } else {
                textViewServer.setTextAppearance(R.style.TextSmall)
            }
            true
        } else {
            textViewServer.visibility = View.GONE
            false
        }
    }

    protected var isNeedUpdateApp = false
    protected var isNeedUpdateLibrary = false
    protected var isNeedUpdateLanguages = false
    protected var isNeedUpdateBorders = false
    protected var isNeedUpdateLocalConfig = false

    fun fillVersionsContext(versions: Versions, versionsServer: Versions) {
        isNeedUpdateApp = checkAndSetItem(softversionBinding!!.app, versions.app!!, softversionBinding!!.appNew, versionsServer.app, true) == true
        isNeedUpdateLibrary = checkAndSetItem(softversionBinding!!.lib, versions.library!!, softversionBinding!!.libNew, versionsServer.library, false) == true
        isNeedUpdateLanguages = checkAndSetItem(softversionBinding!!.lang, versions.languages!!, softversionBinding!!.langNew, versionsServer.languages, false) == true
        isNeedUpdateBorders = checkAndSetItem(softversionBinding!!.border, versions.myAvailableBorders!!, softversionBinding!!.borderNew, versionsServer.myAvailableBorders, false)
        isNeedUpdateLocalConfig = checkAndSetItem(softversionBinding!!.local, versions.localConfig!!, softversionBinding!!.localNew, versionsServer.localConfig, false)
        val v = view
        v?.invalidate()
    }

    fun _fillVersionData() {
        fillVersionsContext(ConfigPrefHelper.getVersions(), ConfigPrefHelper.getVersionsFromServer())
    }

    override fun onErrorCancelClick() {
        requireActivity().finish()
        InitActivity.start(requireContext())
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor?> {
        return CursorLoader(requireContext(), DataContentProvider.TICKETS_URI, arrayOf("COUNT(*)"),
            null, null, null)
    }

    @SuppressLint("SetTextI18n")
    override fun onLoadFinished(loader: Loader<Cursor?>, cursor: Cursor?) {
        var codeCount = 0
        if (cursor != null && cursor.moveToNext()) {
            codeCount = cursor.getInt(0)
        }
        softversionBinding!!.softVer.text = codeCount.toString() // + " " + codeCountText);
    }

    override fun onLoaderReset(loader: Loader<Cursor?>) {}
    override fun onDestroy() {  // TODO Deprecated
        requireActivity().supportLoaderManager.destroyLoader(0)
        super.onDestroy()
    }

    companion object {
        const val TAG = "SoftwareVersions"
        const val LOADER_ID = 225
        fun builder(): Builder {
            return Builder()
        }
    }
}